UOP1_Project/Assets/Scripts/StateMachine/Core/StateMachine.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace DeivSky.StateMachine
+{
+	public class StateMachine : MonoBehaviour
+	{
+#if UNITY_EDITOR
+		public string CurrentState;
+		public bool debug;
+#endif
+		[SerializeField] private Scriptables.ScriptableState _initialStateSO = null;
+		[SerializeField] private ScriptableObject[] _scriptableObjects = null;
```

Comment:
Yeah that's sort of what it is. The idea was to be able to override parameters of an Action/Condition.
It's used in the `ChaseAction` in my example. The asset takes in a default `ChaseDataObject` (or `ChaseDataSO`, I'll rename everything to that format)
``` csharp
[CreateAssetMenu(fileName = "Chase", menuName = "State Machines/Tests/Actions/Chase")]
public class ScriptableChaseAction : ScriptableStateAction
{
	[SerializeField] private ChaseDataObject _chaseData = null;

	protected override StateAction CreateAction() => new ChaseAction(_chaseData);
}
```

That would be the default `ChaseDataSO`. However, on `Awake`, the action checks if there's a different one in the `StateMachine`.
``` csharp
```

Reply:
Isn't this a bit like overthinking? To recap: There's already the ability to assign a different action class, and for each "same" action we can customise its parameters by assigning a different SO to it containing different data, right? (correct me if I'm wrong)

Is it needed to also be able to slightly modify the behaviour of a certain action based on which StateMachine is using it...? Isn't it better to just create a different action altogether? I think this feels a bit like overthinking, introducing a lot of complexity and extra methods all over the FSM.

Reply:
Ah, I see it a bit better now. So basically that array allows state machines to have their own unique data per instance, right?

I guess it's useful then. Ignore my previous comments. But I would still name it to suggest that... `_overrideData` or `_uniqueData`?

Reply:
The annoying thing with this implementation is that this data provided via the array is not tied to the state/action/condition it belongs to. So for instance, having 2 SOs of the same type in this array means you don't know which one is used in the end.
On the other hand if you have 2 state/action/condition that require the same type of data, you have no option to provide 2 different SOs to these... they will both pick up the same one from the array.

A great way would be that whenever you add a state/action/condition to the state machine, this surfaces the field to feed the data (SO) it needs to work to the Inspector of the StateMachine component. In this respect, the Inspector would end up looking a lot like the one Neonage created in #51, but without the logic.

Reply:
One final comment: I feel like you were somehow aware of this issue while coding the whole thing and in fact you broke the pattern you set with the `HungerComponent.cs`, which acts as a container of data unique to the instance of the StateMachine, in fact it's the only non-ScriptableObject script in the StateMachineTest folder.

I guess in a way it shows the flexibility of your implementation (or of Unity? :D) but at the same time it can be confusing: when should I use the array of overrides (`_scriptableObjects`) and when instead use a MonoBehaviour and just dump the data in there?

Reply:
>The annoying thing with this implementation is that this data provided via the array is not tied to the state/action/condition it belongs to. So for instance, having 2 SOs of the same type in this array means you don't know which one is used in the end.
On the other hand if you have 2 state/action/condition that require the same type of data, you have no option to provide 2 different SOs to these... they will both pick up the same one from the array.

I guess I thought, if multiple action/condition use the same type of data, it's likely because they share it, and my idea was to somehow restrict adding two SOs of the same type. Now that you say it though, that's probably not always the case. Unless we make different types for everything, but that wouldn't make much sense.

>A great way would be that whenever you add a state/action/condition to the state machine, this surfaces the field to feed the data (SO) it needs to work to the Inspector of the StateMachine component. In this respect, the Inspector would end up looking a lot like the one Neonage created in #51, but without the logic.

Definitely. I'm really bad at Editor scripting and as I said it was out of scope, but that would be a great way to handle it.

>One final comment: I feel like you were somehow aware of this issue while coding the whole thing and in fact you broke the pattern you set with the HungerComponent.cs, which acts as a container of data unique to the instance of the StateMachine, in fact it's the only non-ScriptableObject script in the StateMachineTest folder.

Yup, you got me! I had to access `HungerComponent` from `IsHungryCondition` and `EnableHungerAction`, also it had to be per instance data, so using SOs made no sense. I was going to overcomplicate it even further by making `HungerComponent` a regular C# class and having another dictionary in the state machine, `Dictionary<Type, object>`, and a `public T GetOrAddObject<T>() where T : class, new()`, this way we could have objects shared between SOs... 
I know, you weren't wrong, I overthink too much :'D. I scraped that and opted for a MonoBehaviour. 

>I guess in a way it shows the flexibility of your implementation (or of Unity? :D) 

I thought about that too while doing that!

>but at the same time it can be confusing: when should I use the array of overrides (_scriptableObjects) and when instead use a MonoBehaviour and just dump the data in there?

I'm trying to get an answer but I seem to be leaning towards removing the array of overrides and just using MonoBehaviours. The array adds an unnecessary layer of complexity, and everything done with it could easily be made with MBs. I think my main problem was that I was thinking of SOs as replacers of MBs, that's why I had exposed values in some actions/conditions. But SOs should simply call the methods in MBs, leaving it up to them however they get configured.

Reply:
This has been really insightful, thank you for all your input!

Reply:
No problem. Do you have an idea of how to improve it then? I would like to merge the changes if we get to a good solution.

Reply:
In terms of the implementation itself, I think just removing the 'override' feature would do. Everything else is specific to the test scene. 
Then a custom editor is definitely needed, and I'd also like to change how transitions are made, maybe have transition tables like kcastagnini in his [#76](https://github.com/UnityTechnologies/open-project-1/pull/76), but a lot of that comes from editor scripting as well and, as I said, I'm not that good at it, so I'd need some help with that.

Reply:
Alright, I removed the `_overrideData` and I'm liking it better now, it's a lot cleaner. I think it's in a decent state now, do you see any other improvements -not related to the inspector- I could work on?